### PR TITLE
Add type hints and update documentation on Models

### PR DESCRIPTION
**Pull request by sdhiscocks**

### CODE REVIEW

**Review by jswright-dstl — APPROVED** [review]

File: stonesoup/models/base.py [hunk 1]
```diff
@@ -43,16 +93,17 @@ class LinearModel(Model):
...
 
-    def function(self, state, noise=False, **kwargs):
+    def function(self, state: State, noise: Union[bool, np.ndarray] = False,
+                 **kwargs) -> Union[StateVector, StateVectors]:
         """Model linear function :math:`f_k(x(k),w(k)) = F_k(x_k) + w_k`
 
         Parameters
         ----------
-        state: :class:`~.State`
+        state: State
```
Comment (oharrald-Dstl) [inline stonesoup/models/base.py:106]: why is this changed?

**Review by oharrald-Dstl — APPROVED** [review]

File: stonesoup/models/base.py (hunk 1 above)
Comment (sdhiscocks) [reply stonesoup/models/base.py:106]: NumpyDoc assumes it's a class when no spaces, so extra `:class:` wrapper is redundant.
